wgpu-hal/src/metal/command.rs

```diff
@@ -21,6 +21,26 @@ impl Default for super::CommandState {
     }
 }
 
+impl Drop for super::CommandState {
+    fn drop(&mut self) {
+        // Metal asserts when a MTLCommandEncoder is deallocated without a call
+        // to endEncoding. This isn't documented at
+        // https://developer.apple.com/documentation/metal/mtlcommandencoder?language=objc
+        // but it manifests as a crash with the message 'Command encoder released without
+        // endEncoding'. To prevent this, we explicitiy call end_encoding on any
+        // encoders we are tracking. Redundant calls appear to be harmless.
+        if let Some(ref encoder) = self.blit {
+            encoder.end_encoding();
+        }
+        if let Some(ref encoder) = self.render {
+            encoder.end_encoding();
+        }
+        if let Some(ref encoder) = self.compute {
+            encoder.end_encoding();
+        }
```

Comment:
question(blocking): Is there a reason that this `Drop` isn't implemented for the individual `{Blit,Render,Compute}Encoder` types themselves?

Reply:
Those types are defined by the `metal` crate. I think the usage rule that encoders must have `endEncoding` called before they are released falls under the category of things that `metal`'s users should be responsible for, not things that `metal` itself should track and take care of internally.